src/metadata/metadata.rs

```diff
@@ -0,0 +1,207 @@
+            }
+            "Level" => {
+                let val = parse::parse_level(val);
+                self.level.join(Some(val));
+            }
+            "Shortname" => {
+                let val = val.clone();
+                self.shortname.join(Some(val))
+            }
+            "Status" => {
+                let val = val.clone();
+                self.raw_status.join(Some(val));
+            }
+            "Title" => {
+                let val = val.clone();
+                self.title.join(Some(val));
+            }
+            _ => die!("Unknown metadata key \"{}\".", key; line_num),
+        }
+
+        self.has_metadata = true;
+    }
+
+    pub fn fill_macros(&self, doc: &mut Spec) {
+        let macros = &mut doc.macros;
+
+        if let Some(date) = self.date.as_ref() {
+            macros.insert(
+                "date",
+                date.format(&format!("{} %B %Y", date.day())).to_string(),
+            );
+            macros.insert("isodate", date.to_string());
+        }
+        if let Some(level) = self.level.as_ref() {
+            macros.insert("level", level.clone());
+        }
+        if let Some(shortname) = self.shortname.as_ref() {
+            macros.insert("shortname", shortname.clone());
+        }
+        if let Some(raw_status) = self.raw_status.as_ref() {
+            macros.insert(
+                "longstatus",
+                SHORT_TO_LONG_STATUS
+                    .get(raw_status.as_str())
+                    .unwrap()
+                    .to_string(),
+            );
+        }
+        if let Some(title) = self.title.as_ref() {
+            macros.insert("title", title.clone());
+            macros.insert("spectitle", title.clone());
+        }
+    }
+
+    pub fn compute_implicit_metadata(&mut self) {
+        if self.canonical_url.is_none() || self.canonical_url.as_ref().unwrap() == "ED" {
+            self.canonical_url = self.ed.clone();
+        }
+    }
+
+    pub fn validate(&self) {
+        if !self.has_metadata {
+            die!("The document requires at least one <pre class=metadata> block.");
+        }
+    }
+}
+
+pub fn parse_metadata(lines: &Vec<Line>) -> (MetadataManager, Vec<Line>) {
+    let mut mm = MetadataManager::new();
+    let mut new_lines: Vec<Line> = Vec::new();
+    let mut in_metadata = false;
+    let mut last_key: Option<String> = None;
+
+    let title_reg = Regex::new(r"\s*<h1[^>]*>(.*?)</h1>").unwrap();
+    let begin_tag_reg = Regex::new(r"<(pre|xmp) [^>]*class=[^>]*metadata[^>]*>").unwrap();
+    let mut end_tag_reg: Option<Regex> = None;
+    let pair_reg = Regex::new(r"([^:]+):\s*(.*)").unwrap();
+
+    for line in lines {
+        if !in_metadata && begin_tag_reg.is_match(&line.text) {
+            // handle begin tag
+            in_metadata = true;
+            mm.has_metadata = true;
+            if line.text.starts_with("<pre") {
+                end_tag_reg = Some(Regex::new(r"</pre>\s*").unwrap());
```

Comment:
Creating a regex is a somewhat expensive operation, so it may be worth it to move it to a local and just keep `end_tag_reg` being an `Option<&Regex>`.

Reply:
The `parse_metadata` would be called many times when we use the [bikeshed watch](https://tabatkins.github.io/bikeshed/#cli-watch) command. What about using the [lazy_static!](https://docs.rs/regex/1.3.7/regex/#example-avoid-compiling-the-same-regex-in-a-loop) to handle that?

Reply:
Yeah, `lazy_static!`-ing the regexes generally makes sense.